addons/mass_mailing_sms/models/mailing_list.py

```diff
@@ -63,3 +63,31 @@ def _mailing_get_opt_out_list_sms(self, mailing):
         opt_out_contacts = subscriptions.filtered(lambda sub: sub.opt_out).mapped('contact_id')
         opt_in_contacts = subscriptions.filtered(lambda sub: not sub.opt_out).mapped('contact_id')
         return list(set(c.id for c in opt_out_contacts if c not in opt_in_contacts))
+
+    def _mailing_list_get_contact_condition(self):
+        email_condition = super(MailingList, self)._mailing_list_get_contact_condition()
+        return "(%s OR %s)" % (
+            email_condition,
+            "contact.phone_sanitized NOT IN (select mobile from phone_blacklist where active = TRUE)"
+        )
+
+    def _mailing_list_get_select_values(self):
+        select_values = super(MailingList, self)._mailing_list_get_select_values()
+        return "%s, %s" % (
+            select_values,
+            "contact.mobile"
+        )
+
+    def _mailing_list_get_rn(self):
+        rn = super(MailingList, self)._mailing_list_get_rn()
+        return "%s, %s" % (
+            rn,
+            "row_number() OVER (PARTITION BY mobile ORDER BY mobile) AS rn_mobile"
+        )
+
+    def _mailing_list_get_rn_where(self):
```

Comment:
not sure to understand this "row number" stuff, why is it needed ?

Reply:
by associating a row number and grouping we can retrieve the first occurence of each email (and discard the rest). When mailing_sms is installed we also get the first occurence of a phone number.

As far as I could tell at least.

Reply:
@std-odoo ``select distinct on`` for the select query seems to have worked fine, I managed to reduce it a decent bit if you want to check it out. 

I applied the other changes (except the commit message which I forgot (will update when I push for the next batch of review)